developer/reference/articles/tooling/blade-cli/08-extending-blade-cli/02-creating-custom-commands-for-blade-cli.markdown

```diff
@@ -0,0 +1,102 @@
+    For example, a `build.gradle` file's `dependencies` block looks like this:
+
+    ```groovy
+    dependencies {
+        compileOnly group: "com.beust", name: "jcommander", version: "1.72"
+        compileOnly group: "com.liferay.blade", name: "com.liferay.blade.cli", version: "latest.release"
+    }
+    ```
+
+3.  Build a Command class by extending the
+    [`BaseCommand`](https://github.com/liferay/liferay-blade-cli/blob/master/cli/src/main/java/com/liferay/blade/cli/command/BaseCommand.java)
+    class:
+
+    ```java
+    import com.liferay.blade.cli.command.BaseCommand;
+
+    public class Hello extends BaseCommand<HelloArgs> {
+
+        @Override
+        public void execute() throws Exception {
+            HelloArgs helloArgs = getArgs();
+
+            getBladeCLI().out("Hello " + helloArgs.getName());
+        }
+
+        @Override
+        public Class<HelloArgs> getArgsClass() {
+            return HelloArgs.class;
+        }
+
+    }
+    ```
+
+    This registers your new command with Blade. The `BaseCommand` class expects
+    an arguments class as its parameter. You'll create this next.
+
+4.  Create a class that holds your command's arguments:
+
+    ```java
+    import com.beust.jcommander.Parameter;
+    import com.beust.jcommander.Parameters;
+
+    import com.liferay.blade.cli.command.BaseArgs;
+
+    @Parameters(commandDescription = "Executes a hello command", commandNames = "hello")
+    public class HelloArgs extends BaseArgs {
+
+        public String getName() {
+            return _name;
+        }
+
+    @Parameter(description = "The name to say hello to", names = "--name", required = true)
+    private String _name;
+
+    }
+    ```
+
+    This class extends the
+    [`BaseArgs`](https://github.com/liferay/liferay-blade-cli/blob/master/cli/src/main/java/com/liferay/blade/cli/command/BaseArgs.java)
+    class. Notice that the class declaration has the `@Parameters` JCommander
+    annotation. This sets your command's description and name. The `@Parameter`
+    annotation applied to the private string `_name` sets how the command's
+    parameter is called and whether it's required.
+
+5.  Since Blade looks for custom commands using the
+    `com.liferay.blade.cli.command.BaseCommand` service interface, you must use
+    a standard JRE service loader mechanism to finish registering your new
+    command with Blade CLI.
+
+    Create a file named `com.liferay.blade.cli.command.BaseCommand` in the
+    `src/main/resources/META-INF/services/` folder. This class should list all
+    of your custom commands' fully qualified class names:
+
+    ```
+    com.liferay.extensions.sample.command.Hello
+    ```
+
+6.  Generate the extension's JAR file (e.g., `gradlew build`).
+
+Awesome! You've created a custom command! See the
+[Installing New Extensions](/docs/7-2/reference/-/knowledge_base/r/installing-new-extensions-for-blade-cli)
+article to install the command (JAR) to Blade CLI.
+
```

Comment:
Should we provide a link somewhere here for the user to read more about ServiceLoader / SPI?
Maybe this one https://www.baeldung.com/java-spi ?

Reply:
Can you provide a little background on this? I can reference that link with a bit of a summary of how we're using the ServiceLoader/SPI.

Reply:
Basically it's a platform for loading classes at runtime, where the file in META-INF/services/ must be named as the interface or base class, and the contents would be one or more FQN's of the implementation classes.

ServiceLoader<T> implements Iterable<T>, T being the interface or base class - matching the name of the file in META-INF/services. So if the classes are available / on the classpath at runtime, the ServiceLoader will return all of the classes it finds implementing or extending T.
The extension classes get on the classpath during blade startup. Blade scans its extensions folder and adds all the extension jars it finds to the classpath - which allows the ServiceLoader to load these extension classes later on.

We do this internally for all our commands in Blade as well.

This might be another good reference link:
https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html